File: First Year/COS110/Assignments/Assignment 4/queue.cpp
#include "queue.h"

template <class T>
Queue<T>::Queue(LinearStructure<T> *c) : OrderedContainer<T>(c) {
    this->front = 0;
    this->rear = 0;
}

template <class T>
Queue<T> & Queue<T>::operator=(const Queue<T> &other) {
    Queue<T> * newQueue = new Queue<T>(this->dataStructure);
    return *newQueue;
}

template <class T>
Queue<T>::Queue(const Queue<T> &other) : OrderedContainer<T>(other.dataStructure) {
    this->front = other.front;
    this->rear = other.rear;
}

template <class T>
Queue<T>::~Queue() {
    delete this->dataStructure;
    this->rear = 0;
    this->front = 0;
}

template <class T>
T Queue<T>::remove() {
    T ret = this->dataStructure->remove(0);
    this->front--;
    return ret;
}

template <class T>
T Queue<T>::next() {
    return this->dataStructure->get(0);
}

template <class T>
void Queue<T>::insert(T el) {
    this->dataStructure->insert(this->front, el);
    this->front++;
}

template <class T>
void Queue<T>::reverse() {

    T * temp = new T[this->front + 1];
    int tFront = this->front;

    int counter = 0;
    while (!this->isEmpty()) {
        temp[counter] = this->remove();
        counter++;
    }

    /*
    cout << "//////////" << endl;
    cout << "COPIED VALS = ";
    for (int i = 0; i < tFront; i++) {
        cout << temp[i] << "\t";
    }
    cout << endl;
    cout << "//////////" << endl;
     */

    Queue<T> * newQueue = new Queue<T>(this->getImplementation());
    newQueue->dataStructure->clear();
    for (int i = tFront - 1; i >= 0; i--) {
        newQueue->insert(temp[i]);
    }
    this->dataStructure = newQueue->getImplementation();
    this->front = tFront;

}

File: Second Year/COS212/Assignments/07_01_Assignment 4/Assignment 1_JAVA_FILES/Queue.java
public class Queue<T extends Comparable<T>> extends OrderedContainer<T> {

    public Queue(LinearStructure<T> c){
        super(c);
    }

    @Override
    public Queue<T> clone() {
        return new Queue<T>(this.dataStructure);
    }

    public Queue(Queue<T> other) {
        super(other.dataStructure);
    }

    public T remove() {
        T temp = null;
        try {
            temp = this.getImplementation().remove(0);
        } catch (RemoveException e) {
            e.printStackTrace();
        }
        return temp;
    }

    public T next() {
        T temp = null;
        try {
            temp = this.dataStructure.remove(0);
            this.dataStructure.insert(0, temp);
        }
        catch (RemoveException ignored) {

        }
        return temp;
    }

    public void insert(T el) {
        //Get back:
        int pos = 0;
        while (true) {
            try {
                T temp = this.dataStructure.remove(pos);
                this.dataStructure.insert(pos, temp);
                pos++;
            }
            catch (RemoveException e) {
                break;
            }

        }
        try {
            this.dataStructure.insert(pos, el);
        } catch (RemoveException ignored) {

        }
    }

}